test/integration/test_server.go

```diff
@@ -0,0 +1,124 @@
+// +build integration,!no-etcd
+
+package integration
+
+import (
+	"fmt"
+	"net/http/httptest"
+	"os"
+	"path"
+	"time"
+
+	kclient "github.com/GoogleCloudPlatform/kubernetes/pkg/client"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/labels"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/util"
+
+	"github.com/openshift/origin/pkg/client"
+	newproject "github.com/openshift/origin/pkg/cmd/experimental/project"
+	start "github.com/openshift/origin/pkg/cmd/server"
+	cmdutil "github.com/openshift/origin/pkg/cmd/util"
+	"github.com/openshift/origin/pkg/cmd/util/tokencmd"
+)
+
+func init() {
+	requireEtcd()
+}
+
+func StartTestServer(args ...string) (start.Config, error) {
+	deleteAllEtcdKeys()
+
+	startConfig := start.NewDefaultConfig()
+
+	basedir := path.Join(os.TempDir(), "openshift-integration-tests")
+
+	startConfig.VolumeDir = path.Join(basedir, "volume")
+	startConfig.EtcdDir = path.Join(basedir, "etcd")
+	startConfig.CertDir = path.Join(basedir, "cert")
+
+	masterAddr := httptest.NewUnstartedServer(nil).Listener.Addr().String()
+	fmt.Printf("masterAddr: %#v\n", masterAddr)
+
+	startConfig.MasterAddr.Set(masterAddr)
+	startConfig.BindAddr.Set(masterAddr)
+	startConfig.EtcdAddr.Set(getEtcdURL())
+
+	startConfig.Complete(args)
+
+	go func() {
+		err := startConfig.Start(args)
+		if err != nil {
+			fmt.Printf("ERROR STARTING SERVER! %v", err)
+		}
+	}()
+
+	// wait for the server to come up
+	if err := cmdutil.WaitForSuccessfulDial(true, "tcp", masterAddr, 100*time.Millisecond, 100*time.Millisecond, 100); err != nil {
+		return *startConfig, err
+	}
+
+	stopChannel := make(chan struct{})
```

Comment:
we definitely want a timeout here... an integration test should just fail within a reasonable amount of time, not hang